waltz-data/src/main/java/com/khartec/waltz/data/survey/SurveyQuestionResponseDao.java

```diff
@@ -122,29 +150,80 @@ public void saveResponse(SurveyInstanceQuestionResponse response) {
+                .select(when(responseExistsCondition, true)
+                        .otherwise(false))
                 .fetchOne(Record1::value1);
 
-        if (responseExists) {
-            dsl.executeUpdate(record);
-        } else {
-            dsl.executeInsert(record);
-        }
+        dsl.transaction(configuration -> {
+            DSLContext txDsl = DSL.using(configuration);
+
+            if (responseExists) {
+                txDsl.executeUpdate(record);
+            } else {
+                txDsl.executeInsert(record);
+            }
+
+            // save survey_question_list_response records
+            response.questionResponse().listResponse().ifPresent(lrs -> {
+                txDsl.deleteFrom(SURVEY_QUESTION_LIST_RESPONSE)
+                        .where(SURVEY_QUESTION_LIST_RESPONSE.SURVEY_INSTANCE_ID.eq(response.surveyInstanceId()))
+                        .and(SURVEY_QUESTION_LIST_RESPONSE.QUESTION_ID.eq(response.questionResponse().questionId()))
+                        .execute();
+
+                if (! lrs.isEmpty()) {
+                    List<SurveyQuestionListResponseRecord> listResponses =
+                            IntStream.range(0, lrs.size())
+                            .mapToObj(i -> {
+                                SurveyQuestionListResponseRecord rec = new SurveyQuestionListResponseRecord();
+                                rec.setSurveyInstanceId(response.surveyInstanceId());
+                                rec.setQuestionId(response.questionResponse().questionId());
+                                rec.setResponse(lrs.get(i));
+                                rec.setPosition(i + 1);
+
+                                return rec;
+                            })
+                            .collect(toList());
+
+                    txDsl.batchInsert(listResponses)
+                            .execute();
+                }
+            });
+        });
     }
 
 
-    public int[] cloneResponses(long sourceSurveyInstanceId, long targetSurveyInstanceId) {
-        Result<SurveyQuestionResponseRecord> records = dsl.select(SURVEY_QUESTION_RESPONSE.fields())
+    public void cloneResponses(long sourceSurveyInstanceId, long targetSurveyInstanceId) {
+        List<SurveyQuestionResponseRecord> responseRecords = dsl
+                .select(SURVEY_QUESTION_RESPONSE.fields())
                 .select(entityNameField)
                 .from(SURVEY_QUESTION_RESPONSE)
                 .where(SURVEY_QUESTION_RESPONSE.SURVEY_INSTANCE_ID.eq(sourceSurveyInstanceId))
-                .fetchInto(SURVEY_QUESTION_RESPONSE);
-
-        records.stream()
-                .forEach(r -> r.setSurveyInstanceId(targetSurveyInstanceId));
-
-        return dsl.batchInsert(records)
-                .execute();
+                .fetchInto(SURVEY_QUESTION_RESPONSE)
+                .stream()
+                .peek(r -> r.setSurveyInstanceId(targetSurveyInstanceId))
+                .peek(r -> r.changed(true))
```

Comment:
I thought using `peek` was mostly to support debug/visibility and using it to modify elements in a stream was generally frowned upon.  

Is this not the same as `.map(r -> { r.setSurveyInstance(t); r.changed(true); return r; })`  ?

Reply:
I used map initially, but IntelliJ suggested peek, but I think I’ll change back to use .map or .forEach.